analysis_service/base/models.py

```diff
@@ -0,0 +1,206 @@
+    public_key = models.CharField(max_length=100000)
+    start_date = models.DateTimeField(blank=True, null=True)
+    end_date = models.DateTimeField(blank=True, null=True)
+    created_by = models.ForeignKey(User, related_name='cluster_created_by')
+
+    jobflow_id = models.CharField(max_length=50, blank=True, null=True)
+    most_recent_status = models.CharField(max_length=50, default="UNKNOWN")
+
+    def __str__(self):
+        return "<Cluster {}>".format(self.identifier)
+
+    def __repr__(self):
+        return "<Cluster {} {}>".format(self.identifier, self.size)
+
+    def get_info(self):
+        return provisioning.cluster_info(self.jobflow_id)
+
+    def is_expiring_soon(self):
+        """Returns true if the cluster is expiring in the next hour."""
+        return self.end_date <= datetime.now().replace(tzinfo=UTC) + timedelta(hours=1)
+
+    def update_status(self):
+        """Should be called to update latest cluster status in `self.most_recent_status`."""
+        info = self.get_info()
+        self.most_recent_status = info["state"]
+        return self.most_recent_status
+
+    def update_identifier(self):
+        """Should be called after changing the cluster's identifier, to update the name on AWS."""
+        provisioning.cluster_rename(self.jobflow_id, self.identifier)
+        return self.identifier
+
+    def save(self, *args, **kwargs):
+        """
+        Insert the cluster into the database or update it if already present,
+        spawning the cluster if it's not already spawned.
+        """
+        # actually start the cluster
+        if self.jobflow_id is None:
+            self.jobflow_id = provisioning.cluster_start(
+                self.created_by.email,
+                self.identifier,
+                self.size,
+                self.public_key
+            )
+
+        # set the dates
+        if not self.start_date:
+            self.start_date = datetime.now().replace(tzinfo=UTC)
+        if not self.end_date:
+            # clusters should expire after 1 day
+            self.end_date = datetime.now().replace(tzinfo=UTC) + timedelta(days=1)
+
+        return super(Cluster, self).save(*args, **kwargs)
+
+    def delete(self, *args, **kwargs):
+        """Remove the cluster from the database, shutting down the actual cluster."""
+        if self.jobflow_id is not None:
+            provisioning.cluster_stop(self.jobflow_id)
+
+        return super(Cluster, self).delete(*args, **kwargs)
+
+
+class Worker(models.Model):
+    identifier = models.CharField(max_length=100)
+    public_key = models.CharField(max_length=100000)
+    start_date = models.DateTimeField(blank=True, null=True)
+    end_date = models.DateTimeField(blank=True, null=True)
+    created_by = models.ForeignKey(User, related_name='worker_created_by')
+    instance_id = models.CharField(max_length=50, blank=True, null=True)
+
+    def __str__(self):
+        return "<Worker {}>".format(self.identifier)
```

Comment:
__str__ should return a human-readable version  of the object. In this case it could simply return self.identifier.